### ISSUE TITLE
Thread pool / Threads

### ISSUE THREAD

**Issue by vhirtham** [vhirtham/GDL#1]
## Multiple Queues
Multiple queues instead of one might be handy. OpenGL calls can't be used in parallel. To avoid parallel calls an extra queue for OpenGL calls can be added which locks access until the latest task has finished. A load balance function must redirect free threads to other queues while the OpenGL queue is locked. Same thing can be used to load data from the hard drive. 
Additionally a load balance function can redirect more threads to time critical tasks or subsystems. For example we can use a separate queue for the physics system which needs to run at a certain frame rate. This queue knows the deadline and other necessary status information to judge if it needs more threads or not. If all tasks are finished the queue should be refilled and locked until the time reaches the current deadline. Afterwards the new deadline is set and the queue gets unlocked.

NOTE: 
The additional queues should be added via templates for performance reasons, since there is no point in having a varying number of queues during execution of the program.
Maybe it is also possible to implement the load balance function via template specialization to reduce overhead. This makes only sense, if this function never changes during execution.

Steps:
-  make member `mQueues` an array with size 1 and get everything to work
-  `TryExecuteTaskWait` loops over all queues and tries to get a task from everyone. If we use a `break` after a successful task execution, queues with higher numbers would rarely be accessed. This is also an integrated load balancing function -> All queues the same. Other load balancing can be used by adding threads with customized main loop. UPDATE: The current version uses a break contrary to the previous statement. If one uses multiple queues, load balancing should be manged by custom thread functions anyways.
- `WaitForTasks` should remain private and only be used by `TryExecuteTaskWait` which must always fetch from all queues (previous bullet point). Otherwise you can get the situation, where a submit notifies a single thread that does not get a task from the loop that caused the notification. In the worst case this results in tasks that are concentrated in one queue and the only thread that is notified does not use this one, causing everything to sleep even though there are tasks for other threads. So a custom loop that should be able to wait but is meant to focus on one (ore more) queue should have an internal `while` loop. This runs as long as its main queue has tasks. If not, the while loop is left and 'TryExecuteTaskWait' is called. If the thread is notified again, it first tries to get a task from all queues and restarts its while loop afterwards.
- HasTasks need a general version (all queues) and a specialized version. Maybe HasTasksQueue. 

**Update**

At the current status (commit 4208eeec1fe802870b5e915cb3b235e9c8a13b16) the general idea for a engine with extra physics loop is as follows: 
A extra thread class which is not managed by the thread pool spawns a thread and runs the physics loop / render loop etc. Each loop enques tasks in the thread pool that could be processed by other threads and grabs tasks from the thread pool as long as it can not proceed (because it is waiting for results or needs to keep up a stable loop rate). However, additional thread pool queues are needed for time consuming operations that should not be processed by external threads that run a loop. 

### Non busy waiting 

To keep the thread pool and its thread classes as simple as possible, there is no internal mechanism for threads to go to sleep. However, a sleep mechanism can be added by providing a custom main loop function to each thread that is started. Just remember to wake up all threads before destroying the thread pool or before closing threads, otherwise you would get a deadlock.

## Debugging
Add some kind of logger (singleton - read drdobbs link below before starting - no synchronizations!!!) to the Threadpool that can be activated by a define (or template parameter if the class is going to be templated anyways [number of queues]). It logs the status of each thread and the queues. Use a ring equivalent container type for logging messages to avoid memory issues. Have a look at page 2 of the drdobbs link below.
Only Synchronization: Protect the creation of the logger in the GetSingleton routine.
Think of writing a GDL::Mutex typedef/wrapper for std::mutex. If the debug define is not defined just use a typedef for std::mutex. If it is defined create a wrapper class that updates the logger and then calls the corresponding mutex function.

(Not related to the previous notes on mutexes, but maybe also useable there:)
- Another option to defines may be a debug template <bool TDebug = false>. This way, the additional code can be activated via constexpr if. The benefit would be, that there is no need to recompile the whole library just to switch to debug mode. Instead, both versions are explicitly instantiated in the cpp and can be used like two separate classes. Maybe add a typedef 

**Good Links**
http://www.drdobbs.com/cpp/multithreaded-debugging-techniques/199200938
https://www.scribd.com/doc/3009706/Debugging-Linux-Applications

**Work that needs to be done**

- [x] Thread class that owns thread
- [x] Add to doxygen docu: How are exceptions handled (@brief)
- [x] Add to doxygen docu: No futures to avoid dead locks and other complications. If results are needed, to proceed,  ask a `atomic<bool>` in the  "continue function"  if the results are available. If not submit the "continue function" again.
- [x] multiple queues (as templates!)
- [x] Add functionality for customizable function calls inside the threads while loop (Instead of just grabbing a task).  
- [x] debug logger
- [x] Terminate on deadlock class - tool for tests  that spawns a thread which calls std::terminate if the program does not set a finished condition in a definable amount of time (Or is destroyed before that).
- [x] Write Thread class for threads that are not part of the thread pool --- read above
- [x] ~~Add template bool to switch between custom allocators and std::allocator (default custom)~~ This is manged by the used GDL containers
- [x] ~~Create unit test (std::allocator) --- Only check true if allocator switch is already implemented~~ Checked by travis setup
- [x] ~~Create integration test (custom allocators)~~ One integration test (mentioned in the missing testvases section) should be enough
- [x] ~~redesign exception handling -> GDL::Vector which stores thrown Exceptions~~ is fine as it is
- [x] Create a benchmark
- [x] Add initialization callback to threadpool threads
- [x] Add deinitialization callback to threadpool threads
- [x] Make std::unique_ptr in thread pool compatible with custom allocators. Therfore a GDL version (UniquePtr) should be introduced and a GDL MakeUnique function.
- [x] UniquePtr deleters need to be classes to avoid problems like non default constructible UniquePtr
- [x] Check if deleter needs to call the destructor. Seems so since for tasks the dtor is not called.
- [x] Check number of allocations in ~~unit~~ integration test when previous bullet point is finished
- [x] Think about an alternative to TryExecuteTaskWait. It is not really flexible when used with multiple queues and might also cause deadlocks in rare cases. Maybe it is better to offer a sleep and wake up function on each thread pool thread.
- [x]  Move threadpool thread class to own header / inl
- [x] ~~Add GetThisThread function to thread pool. Returns thread object of the calling thread or throws if it is not present. Protect it with the thread mutex ---> This makes it impossible for the thread to execute this function while the construction of the object is not finished, since the constructing thread pool function locks the same mutex.~~ Have a look at the non busy waiting topic above
- [x] ~~Add sleep and wakeUp function to thread~~ Have a look at the non busy waiting topic above
- [x] ~~Sleep function can't set internal condition to true if the thread is shutting down! --- OR Condition variable checks also shut down condition.~~ Have a look at the non busy waiting topic above
- [x] ~~Add wakeUpAllThreads function to threadPool~~ Have a look at the non busy waiting topic above
- [x] ~~Make all thread functions private that should be used only by the threadpool (declare friend). Necessary since there is a getter.~~ Have a look at the non busy waiting topic above
- [x] Check if thread pool queue runs faster with a spinlock - No major difference in benchmark, but also not slower
- [x] Check all classes if their implementations can be simplified 

**Missing test cases**
- [x] test that that threads who are started first are closed last (FILO) - important if threads can get alternative/additional main loop callbacks
- [x] test for exception handling
- [x] check thread spawning if implemented
- [x] Write a (integration) test ~~that basically simulates a game engine with render loop, physics loop and a thread pool~~. The thread pool should use multiple queues ~~to separate expensive tasks from simple ones in order to protect external threads that run time critical loops and grab tasks if they have the time.~~ --- Just use multiple queues and multiple memory systems
- [x] Test initialization callback of threadpool threads
- [x] Test deinitialization callback of threadpool threads
- [x] use additional container types which use the general purpose memory and the memory pool to integration test
- [x] ~~Test sleep and wake up functions~~ Have a look at the non busy waiting topic above

**Fix bugs**
- [x] ~~Start_and_Close_Threads test can cause a deadlock. Find origin and fix it (clang-release -> Dev Tools on)~~ Should be fixed by the removal of TryExecuteTaskWait and the condition variable

**Comment by vhirtham** [comment]
Finished with commit 4d7081a42ed7c6804e9db0609464b0f6e31a813b
